### DISCUSSION TITLE
Packaged Mac app crashes on startup

### DISCUSSION THREAD

**Discussion by AndyCW** [unoplatform/uno#14660] [Q&A] [answered]
I have just created my first Uno app - actually a port of an old WPF app - which now runs on Windows, MacOS and Linux, which is great. The app is a development tool as it simulates some medical hardware that we are building. Many of my colleagues use Macs and having this tool running on Macs will help them greatly.
The Mac version of my app runs fine on my machine in Visual Studio for Mac. However, I now want to package the app so that my colleagues can just install the pkg and run it without having to install all the dev platforms and tools just to run this app. 

I am a beginner with Mac app development so it has been a steep learning curve. I have been following the instructions for packaging a MAUI app given here: [Publish a Mac Catalyst app for distribution outside the Mac App Store](https://learn.microsoft.com/en-us/dotnet/maui/mac-catalyst/deployment/publish-outside-app-store?view=net-maui-8.0 ), so I have created an developer ID application cert and an installer cert and a distribution profile, and I have defined build properties in my csproj, similar to this:

```
<PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|net8.0-maccatalyst|AnyCPU'">
  <MtouchLink>SdkOnly</MtouchLink>
  <EnableCodeSigning>True</EnableCodeSigning>
  <EnablePackageSigning>true</EnablePackageSigning>
  <CreatePackage>true</CreatePackage>
  <CodesignKey>Developer ID Application: John Smith (AY2GDE9QM7)</CodesignKey>
  <CodesignProvision>MyMauiApp (Non-App Store)</CodesignProvision>
  <CodesignEntitlements>Platforms\MacCatalyst\Entitlements.plist</CodesignEntitlements>
  <PackageSigningKey>Developer ID Installer: John Smith (AY2GDE9QM7)</PackageSigningKey>
  <UseHardenedRuntime>true</UseHardenedRuntime>
</PropertyGroup>
```

Then I am creating an app package from the command line ` dotnet publish -f net7.0-maccatalyst -c Release -p:CreatePackage=true`. This successfully creates a .pkg for my app. 

Next I send the package to the Apple Notary service, as described here: [Notarize your app](https://learn.microsoft.com/en-us/dotnet/maui/mac-catalyst/deployment/publish-outside-app-store?view=net-maui-8.0#notarize-your-app). This returns a status of Accepted. Finally I staple the ticket to my app and validate. All good.

So now I come to install it. And I share here for anyone reading this who is a newbie like myself, this blocked me for ages. I double-click on the .pkg which starts the installer dialog, run through that, and it reports that it is installed successfully. But the app is nowhere to be seen! It didn't appear in the Applications folder in Finder, nor show up on Launchpad. Baffling. Eventually, I figured out that you must copy the pkg out of the /bin/Release..../Publish folder where it was created, then delete the /bin/Release folder. And it also helps to do a sudo pkgutil --forget <package_id>   - this post is a great help here: https://stackoverflow.com/questions/55516547/install-app-from-pkg-file-is-not-visible-in-application .

OK, so having resolved that, double clicking on the pkg does install the app, and the new app is visible in the Applications folder. Great. But, when I click it to start the app, the icon bounces in the doc for a few seconds and then disappears.

A bit of digging led me to learn about the Console app, where you can get crash dumps for failing apps. My app is crashing with the following (abbreviated):
```
Process:               McStubUno.Mobile [5683]
Path:                  /Applications/McStubUno.Mobile.app/Contents/MacOS/McStubUno.Mobile
Identifier:            com.onward.McStubUno
Version:               1.0.1 (1.0.1)
Code Type:             X86-64 (Translated)
Parent Process:        launchd [1]
User ID:               501
 
Date/Time:             2023-12-02 11:42:08.8184 +0100
OS Version:            macOS 14.1.1 (23B81)
Report Version:        12
Anonymous UUID:        0DA1B80A-9ABC-4749-70A0-F43F473EF7AC
 
Sleep/Wake UUID:       F6EE4740-451C-492C-A1A7-E6621DB21C40
 
Time Awake Since Boot: 6300 seconds
Time Since Wake:       1819 seconds
 
System Integrity Protection: enabled
 
Notes:
PC register does not match crashing frame (0x0 vs 0x7FF8A1802A78)
 
Crashed Thread:        0  tid_103  Dispatch queue: com.apple.main-thread
 
Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
 
Termination Reason:    Namespace SIGNAL, Code 6 Abort trap: 6
Terminating Process:   McStubUno.Mobile [5683]
 
Application Specific Information:
abort() called
 
 
Error Formulating Crash Report:
PC register does not match crashing frame (0x0 vs 0x7FF8A1802A78)
...
```

I have no idea how to resolve this. Is there anyone in the community can help me diagnose what the problem is, and get my app packaged successfully?

[Mac-crashdump.txt](https://github.com/unoplatform/uno/files/13536124/Mac-crashdump.txt)

**Comment by ajpinedam** [comment]
Hey @AndyCW , which Mac do you have, Intel, M1, M2 etc? 

Also, from the crash report, I see you are using Sonoma (macOS 14.1.1). Can you confirm that?

**Reply by AndyCW** [reply]
It is an M2 Pro, and yes, running Sonoma 14.1.1

**Reply by spouliot** [reply]
Bonjour @AndyCW 

We might need more information, like the `.binlog`, to determine what's going on. However I spotted a few things.

# Rosetta

```
Code Type:             X86-64 (Translated)
```

This means you have produced an `x86_64` binary - which is not optimal for your `arm64` Mac.

It's also might not be what you're running when debugging / running it locally.

That's a rather common issue when publishing, you're using the `Release` configuration but it can be quite different from the (working) `Debug` configuration you have been using locally.

You can try to publish the `Debug` configuration and/or to run the `Release` configuration locally. That might expose the problem being related to the differences between the configuration.

# Linking

`<MtouchLink>SdkOnly</MtouchLink>`

You're linking/trimming in release builds. That's good but make sure you have tested the same setting in the `Debug` configuration.

# Packaging

It would be best to publish a _fat_ application, where both `x86_64` and `arm64` binaries are present. That way the consumer of your package will be running the optimal version for their Macs.

**Reply by AndyCW** [reply]
Thanks for your comments.
I have tried running the Release configuration in Visual Studio - it runs. I have tried creating a published output for the Debug configuration - that crashes in the same way as before.

You say: "It would be best to publish a fat application, where both x86_64 and arm64 binaries are present" . OK, but how? I am using the following ´dotnet publish´ command:   ´´´dotnet publish -f net7.0-maccatalyst -c Release -p:CreatePackage=true´´´.  That framework parameter, net7.0-catalyst, means "Universal" does it not?

And in the documentation, it states:
<html>
<body>

<RuntimeIdentifier> | The runtime identifier (RID) for the project. Release builds of .NET MAUI Mac Catalyst apps default to using maccatalyst-x64 and maccatalyst-arm64 as runtime identifiers, to support universal apps. To support only a single architecture, specify maccatalyst-x64 or maccatalyst-arm64.
-- | --

</body>
</html>

so, since I am not specifying a RuntimeIdentifier, it should default to supporting both x64 and arm64, right?

However, one contradictory piece of evidence, is that the published output is created in my projects /bin/debug/net7.0-maccatalyst/maccatalyst-x64/publish folder even though I haven't set the publish output path anywhere. And - as you have noted -when it crashes, it says that the executable is an x64 exe. 
So, I am confused, how do I create a 'fat' application?

**Reply by spouliot** [reply]
> That framework parameter, net7.0-catalyst, means "Universal" does it not?

Yes, it does sounds logical but you can't assume that, e.g. it would be a bad default for `Debug` since it would double the build time. 

So it was, originally, only `x64` but that was [changed](https://github.com/xamarin/xamarin-macios/issues/15620) a while ago for `Release` builds.

However it seems something (could be Uno) is resetting the RID value to `x64` as you do not have the expected files for both architectures.

Again it would be useful to have the `.binlog` of your `dotnet publish` for analysis, and also one for a normal (working) build.


Finally, to avoid the potential issues wrt code signing (and save some time while iterating), you can try a subset of the process with the following instructions: https://learn.microsoft.com/en-us/dotnet/maui/mac-catalyst/deployment/publish-unsigned?view=net-maui-8.0

**Reply by AndyCW** [reply]
I have switched to unsigned.

Here are three binlog files in this ZIP:
- dotnet build -c Debug -f net7.0-maccatalyst  - Runs OK
- dotnet build -c Release -f net7.0-maccatalyst  - Runs OK
- dotnet publish -c Debug-f net7.0-maccatalyst -p:CreatePackage=true -p:Platform=AnyCPU -p:EnableCodeSigning=false -p:EnablePackageSigning=false -  Installs OK, but crashes on startup 

[binlog.zip](https://github.com/unoplatform/uno/files/13560245/binlog.zip)

**Reply by spouliot** [reply]
hmm... I only see 2 binlogs (both named as `WORKS`) inside the attached .zip file.

**Reply by AndyCW** [reply]
Oops, sorry. Here are all three: 
[binlogs - all.zip](https://github.com/unoplatform/uno/files/13578804/binlogs.-.all.zip)

**Reply by spouliot** [reply]
Thanks for the files!

In your `*.Mobile.csproj` try to comment that line

```xml
<RuntimeIdentifier Condition="'$(RuntimeIdentifier)'==''">maccatalyst-x64</RuntimeIdentifier>
```

That was useful on older dotnet but it now interfere with the new defaults (to both architectures) when publishing app.

This should generate the installer with a fat application. Its location on disk will also be different (since it's not architecture specific).

E.g. `bin/Release/net7.0-maccatalyst/publish/*.Mobile-1.0.pkg`

That might not solve the crash... but the crash should be different (and maybe more helpful) so please attach it here.

**Reply by AndyCW** [reply]
I commented out that line forcing the RuntimeIdentifier to maccatalyst-x64 and, yes, now my dotnet publish command does try to build a fat package. However, the build fails, and I get why it does, because of something about this app I haven't shared before as it didn't seem relevant and I didn't want it to become a distraction, though it is clearly a problem when trying to build this fat package.

The error I get when trying to publish is:

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: /Users/andywigley/Projects/arc-mc-stub-uno/McStubUno/McStubUno.Mobile/bin/Release/net7.0-maccatalyst/maccatalyst-x64/McStubUno.Mobile.app/Contents/MonoBundle/libSystem.IO.Ports.Native.dylib and /Users/andywigley/Projects/arc-mc-stub-uno/McStubUno/McStubUno.Mobile/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/McStubUno.Mobile.app/Contents/MonoBundle/libSystem.IO.Ports.Native.dylib have the same architectures (x86_64) and can't be in the same fat output file
```

This is happening because my app uses System.IO.Ports to communicate over a serial port. But there is a bug currently in the .NET System.IO.Ports implementation which causes it to fail to copy the native library to thge built output so the app fails at runtime with a "System.DllNotFoundException: libSystem.IO.Ports.Native" exception. There is a long discussion about this in this thread: https://github.com/dotnet/runtime/issues/75120 , but you can work around this problem simply by getting a copy of the native library **libSystem.IO.Ports.Native.dylib** and including the library on your project and setting the build Copy to Output Dir action to 'Copy Always'. You can get the native library by downloading the NuGet package, unzipping it and copying from the \runtimes\{os-platform}\native folder. The arm64 version is in the https://www.nuget.org/packages/runtime.osx-arm64.runtime.native.System.IO.Ports/ package, and the x64 version in https://www.nuget.org/packages/runtime.osx-x64.runtime.native.System.IO.Ports.

Currently, I have the x64 version included as build content in my project, and clearly, running the project in VS for Mac builds and runs the project as x64, because the app runs fine. But of course, when trying to publish a fat package, the lack of the arm64 version is a problem, hence the error I'm getting. I guess I could do some fancy conditional inclusion stuff in the csproj so as to copy the correct native library for the architecture currently being built, but that's for the future,

So, I can't currently build a fat package, but I can build separate packages for x64 and for arm64, by including the appropriate native library. We already have a crash report for the x64 version, so I built an arm64 version, using:
```
dotnet publish -f net7.0-maccatalyst -c Release  -p:CreatePackage=true -p:Platform=maccatalyst-arm64 -p:RuntimeIdentifier=maccatalyst-arm64  -p:EnableCodeSigning=false -p:EnablePackageSigning=false –bl
```

I installed it and ran it, and this one crashes as well, but with a different reason, so hopefully this will be more informative?:

[arm64-crash-report.txt](https://github.com/unoplatform/uno/files/13601028/arm64-crash-report.txt)

**Reply by spouliot** [reply]
Thanks for the additional data and sorry for the delay (working part time).

So this seems to be two different/orthogonal issues after all. I've opened an [issue](https://github.com/unoplatform/uno.templates/issues/487) about the fat apps.

Back to the crash itself... The new arm64 crash report is related to code signing so it did not provide more clues.

> Exception Type:        EXC_BAD_ACCESS (SIGKILL (Code Signature Invalid))

Let's give this (crash analysis) approach a last go. Would it be possible to try it again without the `-p:EnableCodeSigning=false -p:EnablePackageSigning=false` ?

Thanks again!

**Reply by AndyCW** [reply, marked as answer]
I found out what the problem was! 
This app defines some internal configuration state on first run, and it writes the state into some XML files. The problem is that it writes those files into the folder where the app is located which is perfectly fine running in debug in Visual Studio, but is forbidden for a packaged app running after installation. The solution is to write the files not into AppContext.BaseDirectory, but somewhere else - I use SpecialFolder.CommonApplicationData (/usr/share on Mac).

It is such a shame that the crash report didn't give a clear exception message, then it would be easy to resolve.

Thank you so much for your help and patience @spouliot !

**Comment by spouliot** [comment]
@AndyCW ah, that makes sense! Glad you got it working 😄 

Still I think the arm64 crash report might have given some clues...

Also you could hook the `AppDomain.CurrentDomain.UnhandledException` to a method that writes the exception to the logs using `ObjCRuntime.Runtime.NSLog(...)` (IIRC). You would then be able to see the exception by using Apple's `Console.app`
